tests/serac_odes.cpp

```diff
@@ -0,0 +1,645 @@
+                                      .lin_solver  = LinearSolver::CG,
+                                      .prec        = {}};
+
+NonlinearSolverOptions nonlinear_options{.rel_tol = 1.0e-12, .abs_tol = 1.0e-12, .max_iter = 10, .print_level = -1};
+
+mfem::DenseMatrix M = []() {
+  mfem::DenseMatrix M(3, 3);
+  M(0, 0) = 2.0;
+  M(0, 1) = 1.0;
+  M(0, 2) = 0.0;
+
+  M(1, 0) = 1.0;
+  M(1, 1) = 4.0;
+  M(1, 2) = 1.0;
+
+  M(2, 0) = 0.0;
+  M(2, 1) = 1.0;
+  M(2, 2) = 2.0;
+  return M;
+}();
+
+mfem::DenseMatrix C = []() {
+  mfem::DenseMatrix C(3, 3);
+  C(0, 0) = 1.0;
+  C(0, 1) = 0.0;
+  C(0, 2) = 0.0;
+
+  C(1, 0) = 0.0;
+  C(1, 1) = 1.0;
+  C(1, 2) = 0.0;
+
+  C(2, 0) = 0.0;
+  C(2, 1) = 0.0;
+  C(2, 2) = 1.0;
+  return C;
+}();
+
+std::function stiffness_linear = [](const mfem::Vector & /*x*/) -> mfem::DenseMatrix {
+  mfem::DenseMatrix K(3, 3);
+  K(0, 0) = 1.0;
+  K(0, 1) = -1.0;
+  K(0, 2) = 0.0;
+
+  K(1, 0) = -1.0;
+  K(1, 1) = 2.0;
+  K(1, 2) = -1.0;
+
+  K(2, 0) = 0.0;
+  K(2, 1) = -1.0;
+  K(2, 2) = 1.0;
+  return K;
+};
+
+std::function internal_force_linear = [](const mfem::Vector& x) -> mfem::Vector { return stiffness_linear(x) * x; };
+
+std::function stiffness_nonlinear = [](const mfem::Vector& x) -> mfem::DenseMatrix {
+  mfem::DenseMatrix K(3, 3);
+  K(0, 0) = 1.0;
+  K(0, 1) = -1.0;
+  K(0, 2) = 0.0;
+
+  K(1, 0) = -1.0;
+  K(1, 1) = 2.0 - 2 * x(1) + x(2);
+  K(1, 2) = -1.0 + x(1);
+
+  K(2, 0) = 0.0;
+  K(2, 1) = -1.0 + x(2);
+  K(2, 2) = 1.0 + x(1) + x(2);
+  return K;
+};
+
+std::function internal_force_nonlinear = [](const mfem::Vector& x) -> mfem::Vector {
+  mfem::Vector f(3);
+  f(0) = x(0) - x(1);
+  f(1) = (2.0 - x(1)) * x(1) + (x(1) - 1.0) * x(2) - x(0);
+  f(2) = x(1) * (x(2) - 1.0) + (1.0 + 0.5 * x(2)) * x(2);
+  return f;
+};
+
+mfem::Vector f_ext = []() {
+  mfem::Vector f(3);
+  f(0) = 1.0;
+  f(1) = 2.0;
+  f(2) = 3.0;
+  return f;
+}();
+
+// continuous constraint with continuous derivative
+const auto sine_wave = [](const mfem::Vector& /*x*/, double t) { return 1.0 + sin(2.0 * M_PI * t); };
+
+// continuous constraint with discontinuous derivative
+const auto triangle_wave = [](const mfem::Vector& /*x*/, double t) {
+  return 1.0 + (2.0 / M_PI) * asin(sin(2.0 * M_PI * t));
+};
+
+enum ode_type
+{
+  LINEAR,
+  NONLINEAR
+};
+
+std::string to_string(ode_type o)
+{
+  if (o == LINEAR) return std::string("linear");
+  if (o == NONLINEAR) return std::string("nonlinear");
+  return "unknown";
+}
+
+enum constraint_type
+{
+  UNCONSTRAINED,
+  SINE_WAVE,
+  TRIANGLE_WAVE
+};
+
+std::string to_string(constraint_type c)
+{
+  if (c == UNCONSTRAINED) return std::string("unconstrained");
+  if (c == SINE_WAVE) return std::string("sine wave");
+  if (c == TRIANGLE_WAVE) return std::string("triangle wave");
+  return "unknown";
+}
+
+std::string to_string(serac::TimestepMethod m)
+{
+  // for first order odes
+  if (m == serac::TimestepMethod::BackwardEuler) return std::string("BackwardEuler");
```

Comment:
Hopefully we'll be able to switch to GCC 9 soon so we can start using `magic_enum`

Reply:
@white238 is there anything other than will and time keeping us from GCC9?

Reply:
I've never used `magic_enum` before, is that included as part of axom?

Reply:
GCC 9 is not available on LC yet.  8.3.1 is the newest.  We can always request they install it.